### DISCUSSION TITLE
Windows Webview2 Runtime

### DISCUSSION THREAD

**Discussion by leaanthony** [wailsapp/wails#736] [Q&A]
Greetings all!

It's great that we can now target webkit on windows via WebView2. What isn't so great is:
  * Not every windows machine has it installed
  * The runtime is 150M compressed

So the big question: Should Wails get involved with the installation of the runtime or should that be a deployment concern for the developer? 

Before answering that question, I'd suggest going through [this thread](https://github.com/MicrosoftEdge/WebView2Feedback/issues/341) where they discuss the issue at length.

One thing that keeps playing on my mind is whether or not Wails should generate an NSIS installer that runs the runtime bootstrapper if none is detected. This is obviously a fair bit of work and a detour from the libraries primary goals so would need help if we decide to do this.

👍 for "let's make a distribution solution"
👎 for "that's a deployment concern for the developer"

Also, please feel free to voice your opinions in comments!

**Comment by goldalworming** [comment]
make distribution solution is great idea, the reason using electron is it's very easy from raw code, build and create installer

**Comment by leaanthony** [comment]
I've written some code that queries if the system has webview2 runtime installed and if not, prompt the user whether they want to install it. If they click yes, then the bootstrapper installer is downloaded and executed. If it installs successfully, the application can continue running. This adds a tiny amount of overhead to the app.

Do you think this is better than generating a full blown installer?

**Comment by leaanthony** [comment]
Here's the current solution:

When you run your app on a machine with no runtime, you get this message:

<img width="360" alt="2021-06-15 07_15_23-Windows 10 x64 - VMware Workstation 16 Player (Non-commercial use only)" src="https://user-images.githubusercontent.com/1943904/121960988-2c854d00-cdaa-11eb-90f9-f88991d04cf6.png">

If you click "Ok", then it will download the bootstrapper and run it:

<img width="706" alt="2021-06-15 07_16_01-Windows 10 x64 - VMware Workstation 16 Player (Non-commercial use only)" src="https://user-images.githubusercontent.com/1943904/121961050-3f981d00-cdaa-11eb-994e-a964fa4f7882.png">

Once it is installed, the app will run:

<img width="687" alt="2021-06-15 07_17_19-Windows 10 x64 - VMware Workstation 16 Player (Non-commercial use only)" src="https://user-images.githubusercontent.com/1943904/121961104-53dc1a00-cdaa-11eb-89d5-b2e41cb6fb4c.png">

It's not 100% what I want and there's a small bug but I think this is a pretty good way to go.

**Reply by svengreb** [reply]
I am divided when it comes to “magic“ solutions because it often also comes with many edge cases and maintenance overhead.

Detecting that the required runtime is available is good when the way it works is 100% reliable. This means that it should not be solved by simply looking up for some files in the filesystem as this can either be customized by the user or can change at any time at Microsofts will. Using a detection by calling platform specific functions should be more stable and ensures that it works even when the system changes.

Automatically downloading binaries/scripts via “custom code“ and execute them is kind of a no-go for any OS. There are many arguments against it that range from security perspectives to maintenance overhead. There will be problems for users with different anti-virus applications and configurations where such actions might be blocked. When the currently logged in user is not allowed to install anything at all it could end up with an error that does not describe the actual problem at all. Next up there is the question of how to manage the WebView2 runtime version that gets downloaded and installed, i.e. changed installer location/URL, unavailable servers or unexpected errors during transfer. I'm not sure if it is possible to install multiple versions of the WebView2 runtime (like these hundreds of C++ packages with different versions when installing games 🙈), but if so the users system might end up in a mess.
On the other side an automated installation process should be fine when there is an official way to define and install dependencies, i.e. in a way like Go modules, `npm`, Gradle and so on. I only have a “throwaway“ Windows system these days and never really looked into the ways Microsoft (or third-party tools) provide to install dependencies, but maybe some of these installer types (MSI, NSI etc.) allow to explicitly define dependencies like WebView2.

Don't get me wrong here, it looks like you've already spend time to play around with a working solution, but I just want to share concerns when it comes to “magic“ in software development. I've some experience with it and in the end it does not really end well or adds large overhead in project maintenance. I always like to stick and refer to the [K.I.S.S. principle][1] and [Unix philosophy][2] for projects when it comes to design and architecture decisions.

Finally I'd also like to suggest another solution: Try to detect if the WebView2 runtime is installed and (like on your prototype) inform the user in some way if not, but simply open the official WebView2 website, using the browser set as default by the user. The user can now decide on their own if they really want to install it to run the application or not. This way also ensures that the latest WebView2 runtime is used and prevents problems when there are changes in Microsofts infrastructure regarding the location of installer binaries or maybe even the introduction of bot protections through captcha mechanisms which might break automated solutions build into Wails.

[1]: https://en.wikipedia.org/wiki/KISS_principle
[2]: https://en.wikipedia.org/wiki/Unix_philosophy

**Reply by leaanthony** [reply]
Great feedback @svengreb! I've opted to provide 4 strategies for handling this and it's entirely up to the developer to decide (or create more than one binary depending on scenarios they want to support):

1. Download - Will ask the user if they wish to download and install the official installer* from Microsoft. If they say yes, it will do just that. Microsoft recommends this strategy.
2. Embed - Will ask the user if they wish to install using an embedded version of the runtime installer*. This makes your binaries about 2M heavier. 
3. Browser - Will notify the user they need to install the runtime and open a browser on the download page.
4. Error - Just tell then it needs the runtime and exit.

You can make your decision at build time.

\* The "installer" is the official WebView2 bootstrapper that downloads and installs the latest version of the runtime most appropriate for your system architecture.

**Reply by svengreb** [reply]
Providing multiple solutions is the perfect way, well done 👍🏼 
So when the installer is just one of these wrappers that get the correct version from the correct location it is also totally fine to embed it.

**Reply by iowayfarer** [reply]
I wonder if at some point it will also be worth considering having a fallback for the [evergreen](https://docs.microsoft.com/en-gb/microsoft-edge/webview2/concepts/distribution#evergreen-distribution-mode) installer where it tries to install via [winget](https://docs.microsoft.com/en-gb/windows/package-manager/) then falls back to the browser, rather than embedding the installer?

**Reply by leaanthony** [reply]
That's basically option 1, but instead of using winget, it just downloads the evergreen installer from Microsoft's permalink.

**Reply by iowayfarer** [reply]
Ah fair enough, thanks for explaning

**Reply by leaanthony** [reply]
I should say that it's also option 2 as well :-)

**Reply by Robert-M-Muench** [reply]
We are trying a simple wails-based tool with option 2 *embed* in a corporate environment. The requester to install WebView shows up, but then nothing happens for some users.

Is it be possible for *normal users* to install WV2, or are any administration rights necessary?

**Reply by leaanthony** [reply]
I think that might be a conversation you need to have with the corporation. Regarding "install on run" scenarios, Microsoft introduced a [breaking change](https://github.com/MicrosoftEdge/WebView2Feedback/issues/2030) meaning we could not detect user installs any more. There's a ticket for this: https://github.com/wailsapp/wails/issues/1177
If you would like to take on this ticket, it would be appreciated.

**Reply by Robert-M-Muench** [reply]
You mention a workaround you have in the WV2 ticket. Is this already included in Wails, or is this #1177

What's the workaround?

**Comment by leaanthony** [comment]
Here's a comparison of the effect on binary size for different webview2 strategies:

Default template with `wails build -production -package -upx -webview2 <strategy>`:

| Strategy |      Filesize   | Description |
| :--------- | :----------------- | :------------ |
| Error | 1,492,992 bytes | Notify user that the runtime is needed and exit |
| Browser | 1,495,040 bytes | Notify the user they need to install the runtime and open the download page in  a browser | 
| Download (Default) | 2,134,528 bytes | Notify user that the runtime needs to be installed and offer to download and run the bootstrapper |
| Embed | 3,369,472 bytes | Notify user that the runtime needs to be installed and offer to install using an embedded bootstrapper |

Perhaps `Browser` should be the default strategy?

**Reply by svengreb** [reply]
`Browser` should be fine, at least Wails is scoped to platforms that run a UI so at least one browser can be expected. If not just fall back to `Error`.

**Comment by nickarellano** [comment]
I noticed a few things on Windows builds:

1. Sometimes the applications load a white screen and after reopening it displays the program.
2. Some keyboard shortcuts like Ctrl + R crash the program (refresh)

Is this a WebView thing? This is all still alpha I'm just more curious what these behaviors are.

**Reply by leaanthony** [reply]
Thanks for raising this! Could you please open an issue with steps to reproduce and your current environment (`wails doctor`) and I'll look at this as a priority 👍

**Reply by nickarellano** [reply]
Will do! You are really on these comments hahah thank you!

**Reply by kylege** [reply]
I have the same situation here

**Reply by leaanthony** [reply]
There were fixes to this pushed to the branch. Please try the latest

**Comment by pobearm** [comment]
My soft is running with offline environment (no internet),  i download MicrosoftEdgeWebView2RuntimeInstallerX64.exe 180MB, how coud i embbed this installer to my app installer?

**Reply by leaanthony** [reply]
`//go:embed` ?
